c10/core/Device.h

```diff
@@ -30,13 +30,34 @@ using DeviceIndex = int16_t;
 struct C10_API Device final {
   using Type = DeviceType;
 
-  /// Constructs a new `Device` from a `DeviceType` and an optional device
+  /// Constructs a new `Device` from a `DeviceType` and a device
   /// index.
-  /* implicit */ Device(DeviceType type, DeviceIndex index = -1)
+  /// Also performs validation
+  /* implicit */ Device(DeviceType type, DeviceIndex index)
       : type_(type), index_(index) {
     validate();
   }
 
+  /// Note: no validation is required in this constructor, since the DeviceIndex
+  /// is set to -1.
+  /* implicit */ Device(DeviceType type) : type_(type), index_(-1) {}
+
+  /// Device specific constructors
+  /// They don't require validating the corresponding DeviceType/DeviceIndex
+  /// combination and are therefore slightly faster.
+
+  static Device hip_unchecked(DeviceIndex index) {
+    Device device(kHIP);
+    device.index_ = index;
+    return device;
+  }
+
+  static Device cuda_unchecked(DeviceIndex index) {
+    Device device(kCUDA);
+    device.index_ = index;
+    return device;
+  }
```

Comment:
There's a different idiom you can use here which is better, see `c10/cuda/CUDAStream.h` and its use of Unchecked.

The reason why this idiom is better is because the unchecked constructors are honest to goodness constructors, which makes them more widely applicable as you can now do unchecked constructions in place, instead of having to allocate a temporary object and then move assigning it into the location. It is also more obviously optimized, though I think the optimizer should be able to handle this code too.

Reply:
Oh, nice. Thanks for pointing that out.